solr/modules/aws-secret-provider/src/test/org/apache/solr/secret/zk/AWSSecretCredentialsProviderTest.java

```diff
@@ -0,0 +1,440 @@
+import static org.apache.solr.secret.zk.AWSSecretManagerCredentialsInjector.SECRET_CREDENTIAL_PROVIDER_SECRET_NAME_VM_PARAM;
+import static org.apache.solr.secret.zk.AWSSecretManagerCredentialsInjector.SecretMultiCredentials;
+
+import java.lang.invoke.MethodHandles;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.solr.SolrTestCaseJ4;
+import org.apache.solr.cloud.AbstractZkTestCase;
+import org.apache.solr.cloud.ZkTestServer;
+import org.apache.solr.common.cloud.DigestZkACLProvider;
+import org.apache.solr.common.cloud.DigestZkCredentialsProvider;
+import org.apache.solr.common.cloud.SecurityAwareZkACLProvider;
+import org.apache.solr.common.cloud.SolrZkClient;
+import org.apache.solr.common.cloud.ZkCredentialsInjector;
+import org.apache.solr.common.cloud.ZkCredentialsInjector.ZkCredential;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException.NoAuthException;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AWSSecretCredentialsProviderTest extends SolrTestCaseJ4 {
+
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private static final Charset DATA_ENCODING = StandardCharsets.UTF_8;
+
+  private static final String ALL_USERNAME = "connectAndAllACLUsername";
+  private static final String ALL_PASSWORD = "connectAndAllACLPassword";
+  private static final String READONLY_USERNAME = "readonlyACLUsername";
+  private static final String READONLY_PASSWORD = "readonlyACLPassword";
+
+  public static final String SECRET_NAME = "zkSecretCredentialSecretName";
+
+  protected ZkTestServer zkServer;
+
+  protected Path zkDir;
+
+  @BeforeClass
+  public static void beforeClass() {
+    System.setProperty("solrcloud.skip.autorecovery", "true");
+  }
+
+  @AfterClass
+  public static void afterClass() {
+    System.clearProperty("solrcloud.skip.autorecovery");
+  }
+
+  @Override
+  public void setUp() throws Exception {
+    super.setUp();
+    assumeWorkingMockito();
+    if (log.isInfoEnabled()) {
+      log.info("####SETUP_START {}", getTestName());
+    }
+
+    zkDir = createTempDir().resolve("zookeeper/server1/data");
+    log.info("ZooKeeper dataDir:{}", zkDir);
+    setSecuritySystemProperties();
+    zkServer = new ZkTestServer(zkDir);
+    zkServer.run(false);
+
+    System.setProperty("zkHost", zkServer.getZkAddress());
+
+    setSecretDigestZkSystemProps();
+    System.setProperty(
+        ZK_CREDENTIALS_INJECTOR_CLASS_NAME_VM_PARAM_NAME,
+        AllAndReadonlyCredentialZkCredentialsInjector.class.getName());
+
+    SolrZkClient zkClient =
```

Comment:
Why so many ZK clients? Can't the one be used over and over for each create? At least most of the zk clients look to be the same. Also use try-with-resources to avoid the explicit need to close the zk clients

Reply:
Hi @risdenk 
This is an existing (copied) code. For some reason (race condition?) you need to close and recreate ZKClient to make the nodes creation successful. You cannot reuse it in the last block because you want to connect to ZK with 'OPEN_ACL_UNSAFE' ACLs.

Refactored the existing and new tests classes to void code duplication. Created a new method _setUpZKNodes_ method.

I think _AbstractDigestZkACLAndCredentialsProvidersTestBase_ (which BTW is not _abstract_) needs more refactoring. Didn't want to make this PR bigger than it already is, but we should consider favoring composition over inheritance here.